graph.cpp

```diff
@@ -141,6 +141,89 @@ vector<set<Node>> Graph::connected_components() {
   return retval;
 }
 
+vector<set<Node>> Graph::parallel_connected_components() {
+#ifdef USE_FBT_F
+  bf->force_flush(); // flush everything in buffertree to make final updates
+#else
+  wq->force_flush();
+#endif
+  GraphWorker::pause_workers(); // wait for the workers to finish applying the updates
+  // after this point all updates have been processed from the buffer tree
+
+  printf("Total number of updates to sketches before CC %lu\n", num_updates.load()); // REMOVE this later
+  update_locked = true; // disallow updating the graph after we run the alg
+  bool modified;
+#ifdef VERIFY_SAMPLES_F
+  GraphVerifier verifier { cum_in };
+#endif
+  pair<Node,Node> query[num_nodes];
+  Node size[num_nodes];
+  vector<Node> reps(num_nodes);
+  fill(size, size + num_nodes, 1);
+  for (Node i = 0; i < num_nodes; ++i) {
+    reps[i] = i;
+  }
+
+  do {
+    modified = false;
+    #pragma omp parallel for default(none) shared(query, reps)
```

Comment:
It seems like we're only running a very small part of this in parallel. Is this the intensive part of the work?

Reply:
It’s the only part that can be efficiently parallel IMO. Since merging deltas requires updating the DSU arrays, they have to be done sequentially.

Reply:
What are the DSU arrays?

Reply:
Disjoint set union; keeps track of the supernodes.

Reply:
So I looked a bit into the theory on this. It seems there is a wait-free version of these disjoint set union structures. However, I don't think we have time before the deadline to make this change.

Reply:
So maybe we can open an issue about this and address it after the deadline.